Dedicated testing doc (#1378)
Napari/napari: docs/source/developers/TESTING.md

We use unit tests, integration tests, and functional tests to ensure that
`napari` works as intended. We have

- Unit tests which test if individual modules or functions work correctly
in isolation.

- Integration tests which test if different modules or functions work properly
when combined.

- Functional tests which test if slices of `napari` functionality work as
intended in the whole system.

To get the most return on investment (ROI) from our coding, we strive to test as
much as we can with unit tests, requiring fewer integration tests, and the least number
of functional tests as depicted in the test pyramid below from
[softwaretestinghelp.com](https://www.softwaretestinghelp.com/the-difference-between-unit-integration-and-functional-testing/):

![tests](images/tests.png)

Unit tests are at the base of the pyramid because they are the easiest to write and
the quickest to run. The time and effort to implement and maintain tests increases
from unit tests to integration and functional tests.

## Test organization

All of `napari` tests are located in folders named `_tests`. We keep our unit
tests located in the individual folders with the modules or functions they are

> testing (e.g. the tests for the `Image` layer are located in a `_tests` along
> side the `image.py` file). Our integration and functional tests are located in

Reviewer: ```suggestion
testing (e.g. the tests for the `Image` layer are located in a `_tests` folder
alongside the `image.py` file). Our integration and functional tests are located in
```

the `napari/_tests` folder at the top of the repository.

We also strive to unit test as much of our model file and utils code independently of
our GUI code. These tests are located in the `napari/layers`, `napari/components`,
and `napari/utils` folders. Our GUI code is tests in the `napari/_tests`, 
`napari/_qt`, `napari/_vispy` folders. The `napari/plugins` folder contains a mix
of tests.

## Running Tests

To run our test suite locally, install test requirements and run pytest as follows:

```sh
pip install -r requirements/test.txt
pytest
```

There are a very small number of tests (<5) that require showing GUI elements, (such
as testing screenshots). By default, these are only run during continuous integration.
If you'd like to include them in local tests, set the environment variable "CI":

```sh

> export CI=1

Reviewer: This could also be:

```suggestion
CI=1 pytest
```

Reviewer 2: good call.  that's probably better

   convenient methods for interacting with your GUI tests (clicking, waiting
   signals, etc...).  See the [`qtbot` docs](https://pytest-qt.readthedocs.io/en/latest/reference.html#pytestqt.qtbot.QtBot) for details.

    ```python
    # the qtbot provides convenience methods like addWidget
    def test_something_else(qtbot):
        widget = QWidget()
        qtbot.addWidget(widget)  # tell qtbot to clean this widget later
        ...
    ```

2. When writing a test that requires a `napari.Viewer` object, we provide our
   own convenient fixture called `viewer_factory` that will take care of
   creating a viewer and cleaning up at the end of the test.  When using this
   function, it is **not** necessary to use a `qtbot` fixture, nor should you do
   any additional cleanup (such as using `qtbot.addWidget` or calling
   `viewer.close()`) at the end of the test.  Duplicate cleanup may cause an
   error.  Use the factory as follows:

    ```python
    # the viewer_factory fixture is defined in napari/conftest.py
    def test_something_with_a_viewer(viewer_factory):
        # viewer factory takes any keyword arguments that napari.Viewer() takes
        view, viewer = viewer_factory()

>         # note, `view` here is just a pointer to viewer.window.qt_viewer

Reviewer: Incidentally, how necessary is that pointer? Maybe we should just return `viewer` and use the pattern `view = viewer.window.qt_viewer` where needed?

Reviewer 2: very unnecessary!  i think it's vestigial from some early test patterns.  I like the idea of just returning the viewer.  I'll open a PR for that and we can change these docs

Author: That PR #1386 looks great to me. How about I merge this PR first and then you can update docs over there?

Reviewer 2: sure

        # do stuff with the viewer, no qtbot or viewer.close() methods needed.
        ...
    ```

> > If you're curious to see the actual `viewer_factory` fixture definition, it's in `napari/conftest.py`

Reviewer: Also incidentally, does conftest.py need to be public as opposed to `_conftest.py`? I was annoyed at my terminal trying to suggest it as an import path the other day...

Reviewer 2: just tested.  `_conftest` does not work.  I think we could however use a `pytest11` entry_point in our setup.cfg instead... should we choose to go in that direction.

Author: I'm going merge this PR as is now and we can revisit this in follow up PRs

> ### Mocking: "Fake it until you make it"

Reviewer: ```suggestion
### Mocking: "Fake it till you make it"
```

Might as well go all-in with colloquialisms. =P
